File: leetcode/js/summary-ranges.js
/**
 * @param {number[]} nums
 * @return {string[]}
 */
var summaryRanges = function (nums) {
  if (nums.length === 0) return nums;

  const result = [];
  let start = null;
  let end = null;

  for (const num of nums) {
    if (start === null) {
      start = num;
      end = num;
      continue;
    }

    if (end + 1 === num) {
      end = num;
    } else {
      if (start === end) {
        result.push(`${start}`);
      } else {
        result.push(`${start}->${end}`);
      }

      start = num;
      end = num;
    }
  }

  if (start === end) {
    result.push(`${start}`);
  } else {
    result.push(`${start}->${end}`);
  }

  return result;
};

var summaryRanges = function (nums) {
  const result = [];
  let start = nums[0];

  for (let i = 1; i <= nums.length; i++) {
    if (nums[i] - nums[i - 1] === 1) continue;

    if (start === nums[i - 1]) {
      result.push(`${start}`);
    } else {
      result.push(`${start}->${nums[i - 1]}`);
    }

    start = nums[i];
  }

  return result;
};

File: leetcode/rust/summary-ranges.rs
impl Solution {
    pub fn summary_ranges(nums: Vec<i32>) -> Vec<String> {
        if nums.is_empty() {
            return vec![];
        }

        let mut result = Vec::new();
        let mut start = nums[0];
        let mut end = nums[0];

        for i in 0..nums.len() {
            if i + 1 < nums.len() && nums[i + 1] == end + 1 {
                end += 1;
            } else {
                let range = if start == end {
                    format!("{}", start)
                } else {
                    format!("{}->{}", start, end)
                };

                result.push(range);

                if i + 1 < nums.len() {
                    start = nums[i + 1];
                    end = nums[i + 1];
                }
            }
        }

        result
    }
}
